NHibernate.Caches.Common.Tests/CacheFixture.cs

```diff
@@ -54,6 +55,79 @@ public void TestRemove()
 			Assert.That(item, Is.Null, "item still exists in cache after remove");
 		}
 
+		[Test]
+		public void TestLockUnlock()
+		{
+			if (!SupportsLocking)
+				Assert.Ignore("Test not supported by provider");
+
+			const string key = "keyTestLock";
+			const string value = "valueLock";
+
+			var cache = GetDefaultCache();
+
+			// add the item
+			cache.Put(key, value);
+
+			cache.Lock(key);
+			Assert.Throws<CacheException>(() => cache.Lock(key));
+
+			Thread.Sleep(cache.Timeout / Timestamper.OneMs);
+
+			for (var i = 0; i < 2; i++)
+			{
+				var lockValue = cache.Lock(key);
+				cache.Unlock(key, lockValue);
+			}
+		}
+
+		[Test]
+		public void TestConcurrentLockUnlock()
+		{
+			if (!SupportsLocking)
+				Assert.Ignore("Test not supported by provider");
+
+			const string value = "value";
+			const string key = "keyToLock";
+
+			var cache = GetDefaultCache();
+
+			cache.Put(key, value);
+			Assert.That(cache.Get(key), Is.EqualTo(value), "Unable to retrieved cached object for key");
+
+			// Simulate NHibernate ReadWriteCache behavior with multiple concurrent threads
+			// Thread 1
+			var lockValue = cache.Lock(key);
+
+			// Thread 2
+			try
+			{
+				Assert.Throws<CacheException>(() => cache.Lock(key), "The key should be locked");
+			}
+			finally
+			{
+				cache.Unlock(key, lockValue);
```

Comment:
I do not think this `Unlock` should be done. It should cause the next assert to fail. And a normal pattern is to unlock only if the locking has succeeded. (Anyway, "thread 2" is not supposed to have a lockValue for unlocking.)

Reply:
Correct, with NH 5.2 unlock won't be called anymore in this scenario. I will modify this and commit along with other changes.

Reply:
Corrected.